pkg/util/kubernetes/apiserver/hpa_controller.go

```diff
@@ -139,30 +148,27 @@ func (c *AutoscalersController) processingLoop() {
 				}
 				c.gc()
 			case <-batchFreq.C:
-				c.toStore.m.Lock()
-				data := c.toStore.data
-				c.toStore.m.Unlock()
-				if !c.le.IsLeader() || len(data) == 0 {
-					continue
-				}
-
-				log.Tracef("Batch call pushing %d metrics", len(data))
-				err := c.pushToGlobalStore(data)
+				err := c.pushToGlobalStore()
 				if err != nil {
 					log.Errorf("Error storing the list of External Metrics to the ConfigMap: %v", err)
-					continue
 				}
 			}
 		}
 	}()
 }
 
-func (h *AutoscalersController) pushToGlobalStore(toPush []custommetrics.ExternalMetricValue) error {
+func (h *AutoscalersController) pushToGlobalStore() error {
 	// reset the batch before submitting to avoid a discrepancy between the global store and the local one
+	localStore := h.toStore
 	h.toStore.m.Lock()
 	h.toStore.data = nil
 	h.toStore.m.Unlock()
-	return h.store.SetExternalMetricValues(toPush)
+	if !h.le.IsLeader() {
+		return nil
+	}
+	log.Tracef("Batch call pushing %d metrics", len(localStore.data))
+	err := h.store.SetExternalMetricValues(localStore.data)
+	return err
```

Comment:
You are accessing the `h.toStore` outside of the lock. Maybe something like this?

```
if !h.le.IsLeader() {
	return nil
}

h.toStore.m.Lock()
defer h.toStore.m.Unlock()

log.Tracef("Batch call pushing %d metrics", len(h.toStore.data))
err := h.store.SetExternalMetricValues(h.toStore.data)

h.toStore.data = nil

return err
```

Reply:
I'll just move the shallow copy inside the lock